### PR TITLE
[GSoC'23]  Add subresultants PRS method using sparse representation

### PR DESCRIPTION
**Pull request by 1e9abhi1e10**
#### References to other Issues or PRs

Related PR: 
https://github.com/sympy/sympy/pull/25278

Related Issues:
#23131 
#20874 

#### Release Notes

NO ENTRY

### CODE REVIEW

**Comment by oscarbenjamin** [comment]
@jksuom what do you think of this?

**Comment by jksuom** [comment]
I'll try to look into this later in the evening.

**Comment by jksuom** [comment]
This seems to be a copy of dup_inner_subresultants so its logic should be correct. Its efficiency should be tested.

**Comment by 1e9abhi1e10** [comment]
closed and reopened to see the effects of benchmarks https://github.com/sympy/sympy_benchmarks/pull/94

File: sympy/polys/rings.py [hunk 1]
```diff
@@ -2868,6 +2868,96 @@ def pexquo(self, g, x=None):
...
+
+        if n < m:
+            f, g = g, f
+            n, m = m, n
+
+        if f == 0:
+            return [0, 0]
+
+        if g == 0:
+            return [f, 1]
```
Comment (oscarbenjamin) [inline sympy/polys/rings.py:2918]: Do the tests cover all of the braches of the code? If not can you add tests covering all branches?

If you install pytest-cov then you can measure coverage like:
```
$ pytest --cov=sympy.polys.rings sympy/polys/tests/test_rings.py --cov-report=html
```
Then you  can open the `htmlcov/index.html` file in a browser to see a web view of the code that is covered.

**Comment by oscarbenjamin** [comment]
In gh-25394 we discussed an example that seems to be slow both with dense and sparse polynomials. It should be profiled with cprofile to see what is slow and to see if some part of the operation can be made faster. In ipython (or isympy) you can profile something like:
```
In [2]: %prun -s cumulative R = A.resultant(B.compose(D))

         18048965 function calls (18048852 primitive calls) in 6.553 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    6.553    6.553 {built-in method builtins.exec}
        1    0.000    0.000    6.553    6.553 <string>:1(<module>)
        1    0.000    0.000    6.540    6.540 polytools.py:2662(resultant)
        1    0.000    0.000    6.539    6.539 polyclasses.py:668(resultant)
        1    0.000    0.000    6.539    6.539 euclidtools.py:770(dmp_resultant)
        1    0.000    0.000    6.539    6.539 euclidtools.py:552(dmp_prs_resultant)
        1    0.002    0.002    6.539    6.539 euclidtools.py:449(dmp_inner_subresultants)
  185/143    0.000    0.000    5.814    0.041 densearith.py:792(dmp_mul)
      179    0.060    0.000    5.811    0.032 densearith.py:735(dup_mul)
        7    0.001    0.000    5.739    0.820 densearith.py:1202(dmp_prem)
       34    0.000    0.000    5.696    0.168 densearith.py:160(dmp_mul_term)
       33    0.000    0.000    5.696    0.173 densearith.py:184(<listcomp>)
    25452    0.074    0.000    4.705    0.000 fields.py:490(__mul__)
    51242    2.107    0.000    3.210    0.000 rings.py:1090(__mul__)
    49487    0.056    0.000    2.757    0.000 fields.py:300(new)
    49487    0.054    0.000    2.647    0.000 rings.py:2219(cancel)
    48760    0.056    0.000    2.408    0.000 rings.py:2140(cofactors)
    48760    0.707    0.000    2.341    0.000 rings.py:2169(_gcd_monom)
    23024    0.054    0.000    1.510    0.000 fields.py:395(__add__)
    48760    0.478    0.000    0.829    0.000 rings.py:2182(<listcomp>)
        6    0.000    0.000    0.690    0.115 euclidtools.py:513(<listcomp>)
       16    0.000    0.000    0.690    0.043 densearith.py:1653(dmp_quo)
       16    0.000    0.000    0.690    0.043 densearith.py:1608(dmp_div)
       16    0.000    0.000    0.690    0.043 densearith.py:1462(dmp_ff_div)
       16    0.006    0.000    0.689    0.043 densearith.py:1410(dup_ff_div)
  4478325    0.578    0.000    0.578    0.000 <string>:1(monomial_mul)

```

**Comment by 1e9abhi1e10** [comment]
``` python
>>> f = -(-k + n + 1)**3*(-k + n + 2)**3
>>> g = (k + 1)**3*(k - n)**3
>>> (p, q), opt = parallel_poly_from_expr((f, g), k, field=True, extension=True)    
>>> A = p.monic()
>>> A = p.monic()
>>> h = Dummy('h')
>>> gens = (k,h,n) 
>>> ring = ZZ[gens]
>>> Ap = ring(A.as_expr())
>>> As = ring(A.as_expr())]
>>> D = Poly(k + h, k, h, domain=opt.domain)
>>> B = q.monic()
>>> BDs = ring(B.compose(D).as_expr())
>>> import time                                      
>>> start = time.time()
>>> print( As.subresultants(BDs, ring(k)))
#output 
>>> end = time.time()
>>> print(end - start)
49.82187867164612

**Comment by oscarbenjamin** [comment]
Does it really take 50 seconds? Do you have gmpy2 installed?

**Comment by 1e9abhi1e10** [comment]
No, I didn't have `gmpy2`.
Actually, I just run the code in the terminal to check the timing of the example.

**Comment by 1e9abhi1e10** [comment]
Sorry, Actually in the above comment I was exploring the result in between the process `start` and `end` without caring about the timing. 
Also, I was confused that this time is in `ms` which is not
I calculated these timings multiple times with no mistakes.
``` python 
>>> start = time.time()
>>> print(As.subresultants(BDs, ring(k)))
#output 
>>> end = time.time()
>>> print(end - start)
2.05230712890625
>>>
>>> start = time.time()
>>> print(As.subresultants(BDs, ring(k)))
#output 
>>> end = time.time()
>>> print(end - start)
2.0685553550720215
>>>
>>> start = time.time()
>>> print(As.subresultants(BDs, ring(k)))
#output 
>>> end = time.time()
>>> print(end - start)
2.0763678550720215
>>>
>>> start = time.time()
>>> print(As.subresultants(BDs, ring(k)))
#output 
>>>>> end = time.time()
>>> print(end - start)
2.0729098320007324
>>> 
>>> start = time.time()
>>> print(As.subresultants(BDs, ring(k)))
#output 
>>> end = time.time()
>>> print(end - start)
1.8831496238708496
```

**Comment by 1e9abhi1e10** [comment]
> Do you have gmpy2 installed?
Yes

**Comment by 1e9abhi1e10** [comment]
Should their is a better way to calculate the timings with my local branch?

**Comment by oscarbenjamin** [comment]
> Should their is a better way to calculate the timings with my local branch?

You can look at the profile output like I showed above. See also the cProfile module in Python.

The point of profiling is not about measuring how long something takes: we want to know *which* parts of the operation take the most time because then you can look at the relevant code and see if it can be made faster. You can try different variations and ways of changing the code to see whether they make things faster or slower.

**Comment by 1e9abhi1e10** [comment]
This is the cprofile output

``` 
>>> cProfile.run("As.subresultants(BDs, ring(k))")
         9111885 function calls (9111884 primitive calls) in 3.271 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.000    0.000 <frozen importlib._bootstrap>:1053(_handle_fromlist)
        1    0.000    0.000    0.000    0.000 <frozen importlib._bootstrap>:404(parent)
        1    0.000    0.000    3.270    3.270 <string>:1(<module>)
     1924    0.001    0.000    0.001    0.000 <string>:1(monomial_div)
  4523738    0.785    0.000    0.785    0.000 <string>:1(monomial_mul)
        7    0.000    0.000    0.000    0.000 <string>:1(monomial_pow)
        3    0.000    0.000    0.000    0.000 basic.py:112(__sympy__)
        4    0.000    0.000    0.000    0.000 basic.py:189(__hash__)
        1    0.000    0.000    0.000    0.000 boolalg.py:345(__bool__)
        1    0.000    0.000    0.000    0.000 boolalg.py:460(<lambda>)
        1    0.000    0.000    0.000    0.000 cache.py:69(wrapper)
        1    0.000    0.000    0.000    0.000 decorators.py:224(_func)
        1    0.000    0.000    0.000    0.000 decorators.py:58(__sympifyit_wrapper)
        1    0.000    0.000    0.000    0.000 dispatcher.py:231(__call__)
        1    0.000    0.000    0.000    0.000 dispatcher.py:232(<listcomp>)
        1    0.000    0.000    0.000    0.000 domain.py:371(new)
        2    0.000    0.000    0.000    0.000 domain.py:379(__call__)
        1    0.000    0.000    0.000    0.000 domain.py:386(convert_from)
     5578    0.001    0.000    0.003    0.000 domain.py:403(convert)
     5579    0.001    0.000    0.001    0.000 domain.py:465(of_type)
        1    0.000    0.000    0.000    0.000 expr.py:373(__lt__)
        3    0.000    0.000    0.000    0.000 inspect.py:606(getmro)
        1    0.000    0.000    0.000    0.000 integerring.py:164(from_ZZ)
        1    0.000    0.000    0.000    0.000 libmpf.py:454(to_float)
        1    0.000    0.000    0.000    0.000 logic.py:151(fuzzy_not)
        2    0.000    0.000    0.000    0.000 numbers.py:1024(__new__)
        1    0.000    0.000    0.000    0.000 numbers.py:1859(__eq__)
        1    0.000    0.000    0.000    0.000 numbers.py:2230(__eq__)
        1    0.000    0.000    0.000    0.000 numbers.py:3098(__bool__)
        1    0.000    0.000    0.000    0.000 numbers.py:3364(__neg__)
        1    0.000    0.000    0.000    0.000 numbers.py:3411(__eq__)
        1    0.000    0.000    0.000    0.000 numbers.py:3472(_eval_evalf)
        1    0.000    0.000    0.000    0.000 numbers.py:3475(evalf)
        1    0.000    0.000    0.000    0.000 numbers.py:3487(__sub__)
        1    0.000    0.000    0.000    0.000 numbers.py:3571(_as_mpf_val)
        2    0.000    0.000    0.000    0.000 numbers.py:3577(__eq__)
        1    0.000    0.000    0.000    0.000 numbers.py:3708(__eq__)
        1    0.000    0.000    0.000    0.000 numbers.py:665(__float__)
        1    0.000    0.000    0.000    0.000 polynomialring.py:43(new)
        1    0.000    0.000    0.000    0.000 polyutils.py:171(_not_a_coeff)
        1    0.000    0.000    0.000    0.000 relational.py:1163(_eval_fuzzy_relation)
        1    0.000    0.000    0.000    0.000 relational.py:1194(_n2)
        1    0.000    0.000    0.000    0.000 relational.py:1206(_eval_is_ge)
        1    0.000    0.000    0.000    0.000 relational.py:1239(is_lt)
        1    0.000    0.000    0.000    0.000 relational.py:1263(is_ge)
        1    0.000    0.000    0.000    0.000 relational.py:808(__new__)
        1    0.000    0.000    0.000    0.000 relational.py:836(_eval_relation)
     2069    0.000    0.000    0.000    0.000 ring.py:26(quo)
       13    0.006    0.000    0.008    0.001 rings.py:1009(__sub__)
       53    1.733    0.033    2.883    0.054 rings.py:1090(__mul__)
       27    0.000    0.000    0.000    0.000 rings.py:1169(__pow__)
        5    0.000    0.000    0.000    0.000 rings.py:1381(_term_div)
     2069    0.005    0.000    0.006    0.000 rings.py:1400(term_div)
        6    0.005    0.001    0.361    0.060 rings.py:1414(div)
        5    0.000    0.000    0.000    0.000 rings.py:1458(<listcomp>)
        5    0.000    0.000    0.000    0.000 rings.py:1462(<listcomp>)
        6    0.000    0.000    0.361    0.060 rings.py:1538(exquo)
     2069    0.002    0.000    0.003    0.000 rings.py:1546(_iadd_monom)
     2069    0.182    0.000    0.297    0.000 rings.py:1590(_iadd_poly_monom)
       35    0.000    0.000    0.002    0.000 rings.py:1627(degree)
       34    0.001    0.000    0.001    0.000 rings.py:1641(<listcomp>)
     2074    0.001    0.000    0.050    0.000 rings.py:1683(leading_expv)
       34    0.000    0.000    0.000    0.000 rings.py:1901(itermonoms)
       46    0.000    0.000    0.000    0.000 rings.py:1905(iterterms)
       27    0.001    0.000    0.010    0.000 rings.py:2534(coeff_wrt)
       27    0.001    0.000    0.001    0.000 rings.py:2578(<listcomp>)
       27    0.001    0.000    0.001    0.000 rings.py:2584(<listcomp>)
        7    0.002    0.000    2.821    0.403 rings.py:2587(prem)
        1    0.002    0.002    3.270    3.270 rings.py:2871(subresultants)
       10    0.000    0.000    0.000    0.000 rings.py:288(__hash__)
        6    0.000    0.000    0.000    0.000 rings.py:291(__eq__)
        5    0.000    0.000    0.000    0.000 rings.py:296(__ne__)
       99    0.000    0.000    0.000    0.000 rings.py:308(zero)
       15    0.000    0.000    0.000    0.000 rings.py:312(one)
     5578    0.002    0.000    0.004    0.000 rings.py:316(domain_new)
      2/1    0.000    0.000    0.000    0.000 rings.py:329(ring_new)
       27    0.002    0.000    0.006    0.000 rings.py:353(from_dict)
        1    0.000    0.000    0.000    0.000 rings.py:370(_rebuild_expr)
        1    0.000    0.000    0.000    0.000 rings.py:373(_rebuild)
        1    0.000    0.000    0.000    0.000 rings.py:393(from_expr)
       70    0.000    0.000    0.000    0.000 rings.py:403(index)
       43    0.001    0.000    0.001    0.000 rings.py:575(new)
     4138    0.001    0.000    0.001    0.000 rings.py:586(__hash__)
       24    0.000    0.000    0.000    0.000 rings.py:597(copy)
       51    0.006    0.000    0.006    0.000 rings.py:666(strip_zero)
        2    0.000    0.000    0.000    0.000 rings.py:672(__eq__)
       19    0.000    0.000    0.001    0.000 rings.py:929(__neg__)
       19    0.000    0.000    0.000    0.000 rings.py:930(<listcomp>)
        1    0.000    0.000    0.000    0.000 rings.py:935(__add__)
        6    0.000    0.000    0.000    0.000 sympify.py:101(sympify)
        4    0.000    0.000    0.000    0.000 sympify.py:502(_sympify)
        6    0.000    0.000    0.000    0.000 {built-in method builtins.all}
        1    0.000    0.000    3.271    3.271 {built-in method builtins.exec}
       10    0.000    0.000    0.000    0.000 {built-in method builtins.getattr}
        6    0.000    0.000    0.000    0.000 {built-in method builtins.hasattr}
       10    0.000    0.000    0.000    0.000 {built-in method builtins.hash}
     5755    0.000    0.000    0.000    0.000 {built-in method builtins.isinstance}
       80    0.000    0.000    0.000    0.000 {built-in method builtins.iter}
       17    0.000    0.000    0.000    0.000 {built-in method builtins.len}
     2108    0.050    0.000    0.050    0.000 {built-in method builtins.max}
       12    0.000    0.000    0.000    0.000 {method 'append' of 'list' objects}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
  4543804    0.476    0.000    0.476    0.000 {method 'get' of 'dict' objects}
        1    0.000    0.000    0.000    0.000 {method 'index' of 'tuple' objects}
     2326    0.000    0.000    0.000    0.000 {method 'items' of 'dict' objects}
       34    0.000    0.000    0.000    0.000 {method 'keys' of 'dict' objects}
        1    0.000    0.000    0.000    0.000 {method 'lower' of 'str' objects}
        1    0.000    0.000    0.000    0.000 {method 'pop' of 'dict' objects}
        1    0.000    0.000    0.000    0.000 {method 'replace' of 'str' objects}
        1    0.000    0.000    0.000    0.000 {method 'rpartition' of 'str' objects}
        2    0.000    0.000    0.000    0.000 {method 'startswith' of 'str' objects}
```

**Comment by oscarbenjamin** [comment]
It is rarely useful to see the profile output without sorting.

Anyway the point is not to dump the profile output here but to look at it and then go look at the code, figure out what is taking the most time and whether it can made faster. Understand why each step takes the time that it does and consider whether it is actually a reasonable amount of time.

Another useful tool for this sort of things is Robert Kern's line_profiler.

**Comment by oscarbenjamin** [comment]
> Another useful tool for this sort of things is Robert Kern's line_profiler.

Also a new one I recently used that is quite good is the scalene profiler.

**Comment by 1e9abhi1e10** [comment]
@oscarbenjamin  It looks good for you?

**Pull request merged**